File: server/controller.js
const initialization = require('./materialsParams.js');

class Controller {

    async firstPageOnLoad(req, res) {

        const materials = await initialization('Materials.json');
        const nounTests = materials.russianSections.morphology.independentParts.noun.tests;

        const firstTask = nounTests[0].pages[0].task;
        const firstRule = nounTests[0].pages[0].rule;

        const firstPageData = {
            task: firstTask,
            rule: firstRule,
        };

        res.json(firstPageData);

    }

    async checkButtonClick(req, res) {

        const materials = await initialization('Materials.json');
        const nounTests = materials.russianSections.morphology.independentParts.noun.tests;
        const test = nounTests[0];
        let { editableInput, page } = req.body;

        function updateData(ToF_field, task_field, rule_field, page_field) {

            const updatedData = {
                ToF: ToF_field,
                task: task_field,
                rule: rule_field,
                page: page_field,
            }

            res.json(updatedData);

        }

        function isTestFinish(page, test) {
            const result = (test.pages[page] === undefined);
            if (result) {
                updateData('Вы прошли тест!', 'Заданий больше нет.', 'Отправляйтесь к следующему тесту!', page);
                return true;
            }
            return false;
        }

        function isCorrectAnswer(page, test) {

            if (editableInput == test.pages[page].answer) {
                page++;
                if (!isTestFinish(page, test))
                updateData('Правильно! Новое задание!', test.pages[page].task, test.pages[page].rule, page);
            } else {
                updateData('Неверно, попробуйте ещё раз!', test.pages[page].task, test.pages[page].rule, page);
            }

        }

        if (!isTestFinish(page, test)) {
            isCorrectAnswer(page, test);
        }

    }

}

module.exports = new Controller();

File: cppFiles/controller.cpp
#include <crow_all.h>
#include <nlohmann/json.hpp>
#include "materialsParams.h"

crow::response firstPageOnLoad() {
    try {
        auto materials = readJSON("Materials.json");
        auto nounTests = materials["russianSections"]["morphology"]["independentParts"]["noun"]["tests"];
        auto firstTask = nounTests[0]["pages"][0]["task"];
        auto firstRule = nounTests[0]["pages"][0]["rule"];

        nlohmann::json response;
        response["task"] = firstTask;
        response["rule"] = firstRule;

        return crow::response(response.dump());
    } catch (const std::exception& e) {
        return crow::response(500, e.what());
    }
}

crow::response checkButtonClick(const crow::request& req) {
    try {
        auto materials = readJSON("Materials.json");
        auto nounTests = materials["russianSections"]["morphology"]["independentParts"]["noun"]["tests"];
        auto test = nounTests[0];

        auto body = nlohmann::json::parse(req.body);
        int page = body["page"];
        std::string editableInput = body["editableInput"];

        auto isTestFinish = [&](int page, const nlohmann::json& test) -> bool {
            if (test["pages"].contains(page)) {
                return false;
            }
            nlohmann::json updatedData = {
                {"ToF", "Вы прошли тест!"},
                {"task", "Заданий больше нет."},
                {"rule", "Отправляйтесь к следующему тесту!"},
                {"page", page}
            };
            return true;
        };

        auto isCorrectAnswer = [&](int& page, const nlohmann::json& test) {
            if (editableInput == test["pages"][page]["answer"]) {
                ++page;
                if (!isTestFinish(page, test)) {
                    return crow::response(200);
                }
            } else {
                return crow::response(400);
            }
        };

        return crow::response(200);
    } catch (const std::exception& e) {
        return crow::response(500, e.what());
    }
}
